[Zern Teo] iP

teozern1 opened the pull request:
> "Wearing a shiny crown, paid by a blank cheque" - Daisuke Ishiwatari

# Eddie: Chatbot
### Basic Features:
1. List tasks
2. Add tasks
3. Remove tasks
4. Mark tasks as done
5. Search for a task using a keyword
6. Clear all tasks in the list

Eddie is your friend which helps you relieve yourself of your mental stack! He is:
- Easy to use
- _Friendly_ 😈 
- 100% **Reliable!** 💯 💯 🔥 

### Features to be added
- [ ] Updated GUI
- [ ] Using calendar to note dates
- [ ] ~~Various~~  Some bug fixes

DOWNLOAD [HERE](https://github.com/teozern1/ip/releases/tag/A-Jar)

### Basic Commands:
* `todo` add a task without a time constraint
* `deadline` add a task with a `/by` date
* `event` add a task with `/from` date and `/to` date
* `bye` to exit the chatbot

```java
    public static void exit() {
        Ui.exit();
        System.exit(0);
    }
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

teozern1 committed:
Added welcome message and exit message

teozern1 committed:
Added exit and echo methods

teozern1 committed:
Changed the command from "exit" to "bye"

teozern1 committed:
Added list and add task methods

teozern1 committed:
Added Task Class, mark, unmark functionality

teozern1 committed:
Added 3 new subclasses for task and the methods required

teozern1 committed:
Added test cases for deadline, todo, event, list, bye

teozern1 committed:
Added exception handling for empty description no /by /from /to

teozern1 committed:
Added delete task function

teozern1 committed:
Add save to hard disk using filewriter

teozern1 committed:
Merge branch 'branch-Level-7'

teozern1 committed:
Ad LocalDate formats

teozern1 committed:
Merge branch 'Branch-Level-8'

teozern1 committed:
Add Storage,TaskList classes

teozern1 committed:
Ignore .txt, Add Parser and Ui classes

teozern1 committed:
Merge branch 'A-MoreOOP'

teozern1 committed:
Package into Eddie

teozern1 committed:
Merge branch 'add-gradle-support'

teozern1 committed:
Add Junit Tests

teozern1 committed:
Merge branch 'Branch-A-JUnit'

teozern1 committed:
Add javadocs

teozern1 committed:
Add find function

teozern1 committed:
Merge branch 'A-Level-9'

teozern1 committed:
Merge branch 'Branch-A-JavaDoc'

# Conflicts:
#	src/main/java/Eddie/Duke.java

teozern1 committed:
Add GUI

jylow commented on src/main/java/Eddie/DialogBox.java:
```diff
@@ -0,0 +1,50 @@
+package Eddie;
+import javafx.geometry.Pos;
+import javafx.scene.control.Label;
+import javafx.scene.image.ImageView;
+import javafx.scene.layout.HBox;
+import javafx.collections.FXCollections;
+import javafx.collections.ObservableList;
+import javafx.scene.Node;
```
Imports should be listed lexicographically. ie. collections -> geometry -> scene

jylow commented on src/main/java/Eddie/Parser.java:
```diff
@@ -0,0 +1,119 @@
+                task.taskIsDone();
+                Storage.write();
+                return Ui.mark(taskNum);
+            } else if (command.equals("unmark")) {
+                int taskNum = sc.nextInt();
+                Task task = TaskList.get(taskNum - 1);
+                task.taskNotDone();
+
+                Storage.write();
+                return Ui.unmark(taskNum);
+            } else if (command.equals("delete")) {
+                int index = sc.nextInt();
+                return Duke.delete(index - 1);
+            } else if (command.equals("todo")) {
+
+                String restOfString = sc.nextLine();
+                if (restOfString.length() != 0) {
+                    String taskName = restOfString;
+                    Task taskToAdd = new Todo(taskName);
+                    return Duke.add(taskToAdd);
+
+                } else {
+                    throw new EmptyDescriptionException();
+                }
```
Small nitpick but is there a reason for leaving empty lines in the middle here?

teozern1 replied:
Oops seems like I deleted something but didnt remove the line

jylow commented on src/main/java/Eddie/Task.java:
```diff
@@ -0,0 +1,46 @@
+
+    public Task(){};
+    public Task(String name) {
+        this.done = false;
```
iirc "done"(booleans) is automatically initialised to false by java compiler so idt you need this statement

jylow commented on src/main/java/Eddie/Ui.java:
```diff
@@ -0,0 +1,73 @@
+     * Prints the message for when Eddie exits.
+     */
+    public static String exit() {
+        String goodbye = "Bye. Hope to see you again soon!";
+        return goodbye;
```
You can just return the string you dont need to initialise a String before returning

jylow commented on src/main/java/Eddie/MissingByDateException.java:
```diff
@@ -0,0 +1,10 @@
+/**
+ * Excpetion when user does not provide by date
+ */
+public class MissingByDateException extends DukeException{
+    public MissingByDateException() {
+        super("Please use /by to include a due date");
+    }
+}
```
Nice way of making the exceptions more specific to each error 😄

jylow commented on src/main/java/Eddie/Storage.java:
```diff
@@ -0,0 +1,91 @@
+     * @throws IOException Thrown if input is not correct.
+     */
+    private static void writeToFile () throws IOException {
+        FileWriter eddieTaskList = new FileWriter("EddieTaskList.txt");
+        for (int i = 0; i < TaskList.size(); i++) {
+            Task t = TaskList.get(i);
+
+            if (t.getType() == "T") {
+                eddieTaskList.write("T , " + t.getStatus() + " , " + t.getName() + " , \n");
+            } else if (t.getType() == "D") {
+                eddieTaskList.write("D , " + t.getStatus() + " , " + t.getName() + " , " + t.getDeadline() + " , \n");
+            } else if (t.getType() == "E") {
+                eddieTaskList.write("E , " + t.getStatus() + " , " + t.getName() + " , " + t.getStartDate() + " , "
+                        + t.getEndDate() + " , \n");
+            }
+        }
+        eddieTaskList.close();
```
Clean way of writing to file. You can just call FileWriter fw btw.

jylow reviewed (commented):
Coding logic is sound and good use of abstraction to make code look cleaner. Maybe just stylistically can be cleaner by making breaklines more consistent. Overall good job!

teozern1 committed:
Add images to GUI

teozern1 committed:
Add assertions

There are several assumptions made when the parser is used, so we should check whether these assumptions hold true during the execution of the program.

teozern1 committed:
Refactor Parser to use switch cases

Using if else statements makes the code difficult to read, switch statements improve readability

teozern1 committed:
Refactor tasks, exceptions and GUI classes into respective folders

All files were in one package and was disorganized, doing this helps with looking for the correct classes/files and groups relevant files together

teozern1 committed:
Refactor commands into respective classes from duke

This makes it easier to understand the code, improves readability

teozern1 committed:
Make parser exceptions nested instead of happy route

Improves readability of code

teozern1 committed:
Add tag functionality to todo tasks

teozern1 committed:
Merge pull request #1 from teozern1/Branch-Assertions

Add assertions

teozern1 committed:
Merge branch 'master' into Branch-CodeQuality

teozern1 committed:
Merge pull request #2 from teozern1/Branch-CodeQuality

Branch code quality

teozern1 committed:
Add tagging to deadline, event

teozern1 committed:
Fix a write to file bug which causes parsing of date to fail

teozern1 committed:
Update Ui.png and README

Add features and guide to README for users.

teozern1 committed:
Clean up code

Use switch cases in the parser to increase readability, abstracted out try and catch block to another function from readFile().

teozern1 committed:
Add descriptions for the command classes

teozern1 committed:
Update name of Ui.png

teozern1 committed:
Add break statements to switch cases

Dates were wrongly parsed, throwing an error without use of break statements.

teozern1 committed:
Refactor duke to be named as Eddie

teozern1 committed:
Add additional exceptions to be checked, add dependencies to build.gradle

teozern1 committed:
Fix mainClass.set to be launcher class

teozern1 committed:
Add exceptions to be caught
